### PR TITLE
have the C++ compiler generate the tables.

### PR DESCRIPTION
**Pull request by lemire**
Instead of having hard-coded tables, we can ask the C++ to generate them.

There is no obvious performance degradation (benchdata, LLVM 14).

cc @ShogunPanda

### CODE REVIEW

**Comment by ShogunPanda** [comment]
To be honest, since I also believe it won't affect performance, I would merge this.

Even if you did it for fun I actually think it improves readability of the code: rather than manually review the code it is easier to spot which elements are set to true and why.

WDYT?

**Comment by lemire** [comment]
I don’t disagree but we need to investigate the CI issues. At least one is unrelated to this PR, and I will examine it soon.

The performance should be identical, but we should check that it is before merging, as an abundance of caution.

**Comment by lemire** [comment]
@anonrig This could be merged.

Note that if you have an old compiler with partial support for C++17, this code might break it. We require genuine C++17 support.

**Comment by lemire** [comment]
@anonrig I don't know how to fix the linter errors. I ran clang-format on the code.

**Comment by lemire** [comment]
@anonrig I think I fixed the linter issue by reverting clang-format to version 14. Indeed, clang-format is version specific.

File: src/helpers.cpp [hunk 1]
```diff
@@ -302,18 +302,19 @@ ada_really_inline size_t find_next_host_delimiter_special(
...
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
+static constexpr std::array<uint8_t, 256> special_host_delimiters =
+    []() constexpr {
+  std::array<uint8_t, 256> result{};
+  for (size_t i = 0; i < 256; i++) {
+    if (i == ':' || i == '/' || i == '[' || i == '\\' || i == '?') {
```
Comment (anonrig) [inline src/helpers.cpp:309]: Recommendation: We should iterate through these values other than 256.

File: src/helpers.cpp [hunk 2]
```diff
@@ -436,18 +437,18 @@ ada_really_inline size_t find_next_host_delimiter(std::string_view view,
...
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
+static constexpr std::array<uint8_t, 256> host_delimiters = []() constexpr {
+  std::array<uint8_t, 256> result{};
+  for (size_t i = 0; i < 256; i++) {
+    if (i == ':' || i == '/' || i == '?' || i == '[') {
```
Comment (anonrig) [inline src/helpers.cpp:443]: Recommendation: Iterating through values rather than index

File: src/helpers.cpp [hunk 3]
```diff
@@ -761,18 +763,18 @@ find_authority_delimiter_special(std::string_view view) noexcept {
...
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
+static constexpr std::array<uint8_t, 256> authority_delimiter = []() constexpr {
+  std::array<uint8_t, 256> result{};
+  for (size_t i = 0; i < 256; i++) {
+    if (i == '@' || i == '/' || i == '?') {
```
Comment (anonrig) [inline src/helpers.cpp:769]: Same comment in here as well

File: src/unicode.cpp [hunk 4]
```diff
@@ -263,55 +248,30 @@ contains_forbidden_domain_code_point_or_upper(const char* input,
...
+    if (c >= '0' && c <= '9') {
+      result[c] = true;
+    } else if (c >= 'a' && c <= 'z') {
+      result[c] = true;
+    } else if (c >= 'A' && c <= 'Z') {
+      result[c] = true;
+    } else if (c == '+' || c == '-' || c == '.') {
+      result[c] = true;
+    } else
+      result[c] = false;
```
Comment (anonrig) [inline src/unicode.cpp:263]: This false seems unnecessary

File: src/unicode.cpp [hunk 5]
```diff
@@ -263,55 +248,30 @@ contains_forbidden_domain_code_point_or_upper(const char* input,
...
-    0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
-    1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
-
-static_assert(sizeof(is_alnum_plus_table) == 256);
+constexpr static std::array<bool, 256> is_alnum_plus_table = []() constexpr {
```
Comment (anonrig) [inline src/unicode.cpp:251]: Can you add comments to this table (and the previous ones) if you don't mind?

**Comment by lemire** [comment]
@anonrig All reasonable comments.

File: src/helpers.cpp (hunk 2 above)
Comment (lemire) [reply src/helpers.cpp:443]: Done.

File: src/helpers.cpp (hunk 3 above)
Comment (lemire) [reply src/helpers.cpp:769]: Done.

File: src/unicode.cpp (hunk 4 above)
Comment (lemire) [reply src/unicode.cpp:263]: Indeed.

**Comment by lemire** [comment]
@anonrig Code updated.

File: src/helpers.cpp [hunk 6]
```diff
@@ -735,18 +728,19 @@ ada_really_inline void strip_trailing_spaces_from_opaque_path(
...
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
-    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
+static constexpr std::array<uint8_t, 256> authority_delimiter_special =
+    []() constexpr {
+  std::array<uint8_t, 256> result{};
+  for (size_t i = 0; i < 256; i++) {
```
Comment (anonrig) [inline src/helpers.cpp:734]: This iteration is not necessary too.

Comment (lemire) [reply src/helpers.cpp:734]: Indeed.

**Comment by lemire** [comment]
@anonrig Might be good now.

**Review by anonrig — APPROVED** [review]

**Review by anonrig — APPROVED** [review]

**Pull request merged**